src/lightning/pytorch/trainer/connectors/callback_connector.py

```diff
@@ -172,7 +172,15 @@ def _attach_model_callbacks(self) -> None:
         model_callbacks = [model_callbacks] if not isinstance(model_callbacks, Sequence) else model_callbacks
         model_callback_types = {type(c) for c in model_callbacks}
         trainer_callback_types = {type(c) for c in trainer.callbacks}
-        override_types = model_callback_types.intersection(trainer_callback_types)
+        # edge case: if an unmodified callback was added, the logic below would filter it
+        trainer_callback_types.discard(Callback)
+        # exclude trainer callbacks of the same class or subclass
+        override_types = set()
+        for m in model_callback_types:
+            for c in trainer_callback_types:
+                if issubclass(m, c):
+                    override_types.add(c)
+                    break
```

Comment:
This break assumes nobody is doing multiple inheritance with our existing callbacks

```python
class Frankenstein(ModelCheckpoint, TQDMProgressBar):
    ...
```

Reply:
This PR fixes  #16990 if I understand correctly. I added it to the description.

Reply:
Oh I didn't know about that issue. There you describe more edge cases that we could want to consider, so maybe it's worth keeping open